Challo/Challo/Modules/Common/Bookings/BookingAPI.swift

```diff
@@ -0,0 +1,118 @@
+    private let touristAPI: TouristAPI
+    private let trailAPI: TrailAPI
+    
+    init(bookingParser: BookingAPIParser,
+         networkManager: NetworkManager,
+         guideAPI: GuideAPI,
+         touristAPI: TouristAPI,
+         trailAPI: TrailAPI) {
+        self.bookingParser = bookingParser
+        self.networkManager = networkManager
+        self.guideAPI = guideAPI
+        self.touristAPI = touristAPI
+        self.trailAPI = trailAPI
+    }
+
+    func getBookingsForTourist(id: UUID, callback: @escaping ([Booking]) -> Void) {
+        let fullURL = url + "/?touristId=\(id.uuidString)"
+        getBookings(url: fullURL, callback: callback)
+    }
+
+    func getBookingsForGuide(id: UUID, callback: @escaping ([Booking]) -> Void) {
+        let fullURL = url + "/?guideId=\(id.uuidString)"
+        getBookings(url: fullURL, callback: callback)
+    }
+    
+    private func getBookings(url: String, callback: @escaping ([Booking]) -> Void) {
+        networkManager.get(url: url,
+                           headers: [String: String]()) { [weak self] response, error in
+            guard let self = self else {
+                return
+            }
+            
+            if error != nil {
+                return
+            }
+            
+            let bookingsJSON = self.bookingParser.extractBookingsJSON(response: response)
+            var bookings = [Booking]()
+            
+            let bookingsGroup = DispatchGroup()
+            bookingsJSON.forEach {
+                self.handleBookingsJSON(json: $0, group: bookingsGroup) { booking in
+                    bookings.append(booking)
+                }
+            }
+            
+            bookingsGroup.notify(queue: DispatchQueue.main) {
+                callback(bookings)
+            }
+        }
+    }
+
+    private func handleBookingsJSON(json: JSON,
+                                    group: DispatchGroup,
+                                    callback: @escaping (Booking) -> Void) {
+        group.enter()
+        var guide: Guide?
+        var tourist: Tourist?
+        var trail: Trail?
+
+        let internalGroup = DispatchGroup()
+        guard let guideID = bookingParser.extractGuideID(json: json),
+              let touristID = bookingParser.extractTouristID(json: json),
+              let trailID = bookingParser.extractTrailID(json: json) else {
+            return
+        }
+        
+        internalGroup.enter()
+        guideAPI.getGuide(guideId: guideID) { retrievedGuide in
+            guide = retrievedGuide
+            internalGroup.leave()
+        }
+        
+        internalGroup.enter()
+        touristAPI.getTourist(userId: touristID) { retrievedTourist in
+            tourist = retrievedTourist
+            internalGroup.leave()
+        }
+    
+        internalGroup.enter()
+        trailAPI.getTrail(trailId: trailID) { retrievedTrail in
+            trail = retrievedTrail
+            internalGroup.leave()
+        }
+        
```

Comment:
Referencing this ah https://www.donnywals.com/sequencing-tasks-with-dispatchgroup/

Reply:
Yeap